kotlinx-coroutines-core/common/src/CancellableContinuationImpl.kt

```diff
@@ -279,13 +331,30 @@ internal open class CancellableContinuationImpl<in T>(
 
     // Unregister from parent job
     private fun disposeParentHandle() {
+        // If instance is reusable, do not detach on every reuse, #releaseInterceptedContinuation
+        // will do it for us in the end
+        if (isReusable()) return
         parentHandle?.let { // volatile read parentHandle (once)
             it.dispose()
             parentHandle = NonDisposableHandle // release it just in case, to aid GC
         }
     }
 
+    /**
+     * Detaches from the parent.
+     * Invariant: used used from [CoroutineDispatcher.releaseInterceptedContinuation] iff [isReusable] is `true`
+     */
+    internal fun detachChild() {
+        val handle = parentHandle
+        handle?.dispose()
+        parentHandle = NonDisposableHandle
+    }
+
     override fun tryResume(value: T, idempotent: Any?): Any? {
+        // Important invariant: if idempotent tryResume is used, then regular resume cannot be used, since the
+        // thread that invokes this tryResume install its descriptor into the pointer to the corresponding node first,
+        // so no other thread can find it and invoke resume.
+        if (idempotent != null) invalidateReusability()
```

Comment:
>Additional advantage is that you can completely remove the whole NON_REUSABLE state, simply resetting DC's state beck to null after cancellation. 

It makes sense, though I expect usages where it makes sense quite rare.
E.g. it works mostly for patterns like
```
hotLoop {
    if (rareCondition) select { ... }
    else channel.receive()
}
```
but I am fine with that.

Please note that reusability is invalidated not only for idempotent resumes but also for cancellation. Currently, there is no test that exposes this need, but take a look at this code pattern:
```
suspendAtomicCancellableCoroutineReusable {
    // Suppose we have already attached to the parent
    it.resume(Unit) 
    // Here comes parent cancellation that is suppressed by `resume`
}

println(isActive) // false, parent is cancelled
// Now suppose we haven't invalidated reusability
suspendAtomicCancellableCoroutineReusable {
    // Here we DO NOT resume and suspend ourselves
} // 1
```

the sanest behavior here is to throw CE in `(1)`. But if we get rid of `NON_REUSABLE`, we have to check parent's cancellation status on each suspension in `checkCompleted`, potentially making it slower and moving the complexity of reusability state machine into `checkCompleted`.

Do I missing something and you've actually considered this behaviour when proposing the idea with removing `NON_REUSABLE`?

Reply:
Consider what happens without `NON_REUSABLE` state (we replaced it with `null` state in code):

```
suspendAtomicCancellableCoroutineReusable {
    // HERE: DC.state == REUSABLE_CLAIMED
    it.resume(Unit) 
    // Here comes parent cancellation that is suppressed by `resume`
    // HERE: DC.state == Throwable
}
// HERE: DC.state == null instead of NON_REUSABLE

println(isActive) // false, parent is cancelled
// Now suppose we haven't invalidated reusability
// But there's no stored continuation in DC.state (DC.state == null)
suspendAtomicCancellableCoroutineReusable {
    // Here we DO NOT resume and suspend ourselves
} // SO: exception is thrown here when we setupCancellation and attach to parent
```

I don't see any problem without `NON_REUSABLE`. The code becomes simpler, less cases to handle to at keep in mind. 

What is especially important is that `DC.isReusable` becomes simpler (`DC.isReusable when DC.state != null`) which makes `CC.isReusable` easier to reason about (it is used it a several scarily non-atomic check-and-act places).

In fact, I've tried the corresponding change in code (getting rid of `NON_REUSABLE`) and it passes all the tests.